File: java/ProjectEuler/Problem25.java
/*
projecteuler.net problem 25.

1000-digit Fibonacci number.

What is the first term in the Fibonacci sequence to contain 1000 digits?
1 = 1, 2 = 1, 3 = 2, 4 = 3, 5 = 5, 6 = 8, 7 = 13
*/

import java.math.BigInteger;

public class Problem25 {

    public static final int DIGITS = 1000;
    
    /** Counts the number of digits of an integer */
    public static int digits(BigInteger n) {
        int count = 0;
        while (n.compareTo(BigInteger.ZERO) > 0) {
            n = n.divide(BigInteger.TEN);
            count++;
        }
        return count;
    }

    /** Generator of fibonacci sequence */
    public static class Fibonacci {
        private BigInteger a, b, count, buffer;

        public Fibonacci() {
            a = BigInteger.ZERO;
            b = BigInteger.ONE;
            count = BigInteger.ZERO;
        }

        /** get the next Fibonacci number */
        public void next() {
            buffer = a; 
            a = a.add(b);
            b = buffer;
            count = count.add(BigInteger.ONE);
        }

        public BigInteger getCount() {
            return count;
        }

        public BigInteger getCurrent() {
            return a;
        }
    }

    public static BigInteger solution() {
        Fibonacci sequence = new Fibonacci();
        while (digits(sequence.getCurrent()) < DIGITS) {
            sequence.next();
        }
        return sequence.getCount();
    }
    
    /** Print solution */
    public static void main(String[] args) {
        System.out.println(solution());
    }
}

File: python/ProjectEuler/problem25.py
"""
projecteuler.net problem 25.

1000-digit Fibonacci number.

What is the first term in the Fibonacci sequence to contain 1000 digits?
1 = 1, 2 = 1, 3 = 2, 4 = 3, 5 = 5, 6 = 8, 7 = 13
"""

import math

LIMIT = 1000


def digits(n):
    """ Returns the number of digits of a positive integer. """
    return int(math.log10(n)) + 1


def fibo():
    """
    Infinite generator for the Fibonacci sequence.
    Returns the tuple (value, sequence position)
    """
    a = 0
    b = 1
    count = 0
    while True:
        a, b = a + b, a
        count += 1
        yield a, count


def solution(limit):
    """
    Finds the first term in the Fibonacci sequence that contains
    <limit> digits.
    """
    for n in fibo():
        if digits(n[0]) >= limit:
            return n[1]


print solution(LIMIT)
